src/mediarithmics/plugins/custom-action/CustomActionBasePlugin.ts

```diff
@@ -0,0 +1,182 @@
+import * as _ from "lodash";
+
+import {
+  CustomActionRequest,
+  CustomActionPluginResponse,
+  CustomAction,
+} from "./CustomActionInterface";
+import { PropertiesWrapper } from "./../common/BasePlugin";
+import { PluginProperty } from "./../../api/core/plugin/PluginPropertyInterface";
+import { BasePlugin } from "../common";
+
+export interface CustomActionBaseInstanceContext {
+  properties: PropertiesWrapper;
+}
+
+export abstract class CustomActionBasePlugin extends BasePlugin {
+  constructor(enableThrottling = false) {
+    super(enableThrottling);
+
+    this.initCustomAction();
+    this.setErrorHandler();
+  }
+
+  /**
+   *
+   * @param customActionId
+   */
+  async fetchCustomAction(customActionId: string): Promise<CustomAction> {
+    const customActionResponse = await super.requestGatewayHelper(
+      "GET",
+      `${this.outboundPlatformUrl}/v1/scenarios_custom_actions/${customActionId}`
+    );
+    this.logger.debug(
+      `Fetched Custom Action: ${customActionId} - ${JSON.stringify(
+        customActionResponse.data
+      )}`
+    );
+    return customActionResponse.data;
+  }
+
+  /**
+   *
+   * @param customActionId
+   */
+  async fetchCustomActionProperties(
+    customActionId: string
+  ): Promise<PluginProperty[]> {
+    const customActionPropertiesResponse = await super.requestGatewayHelper(
+      "GET",
+      `${this.outboundPlatformUrl}/v1/scenarios_custom_actions/${customActionId}/properties`
+    );
+    this.logger.debug(
+      `Fetched Custom Action Properties: ${customActionId} - ${JSON.stringify(
+        customActionPropertiesResponse.data
+      )}`
+    );
+    return customActionPropertiesResponse.data;
+  }
+
+  protected async instanceContextBuilder(
+    instanceId: string
+  ): Promise<CustomActionBaseInstanceContext> {
+    const customActionProps = await this.fetchCustomActionProperties(
+      instanceId
+    );
+
+    const context: CustomActionBaseInstanceContext = {
+      properties: new PropertiesWrapper(customActionProps),
+    };
+
+    return context;
+  }
+
+  /**
+   *
+   * @param request
+   * @param instanceContext
+   */
+  protected abstract onCustomActionCall(
+    request: CustomActionRequest,
+    instanceContext: CustomActionBaseInstanceContext
+  ): Promise<CustomActionPluginResponse>;
+
+  protected async getInstanceContext(
+    instanceId: string
+  ): Promise<CustomActionBaseInstanceContext> {
+    if (!this.pluginCache.get(instanceId)) {
+      this.pluginCache.put(
+        instanceId,
+        this.instanceContextBuilder(instanceId),
+        this.getInstanceContextCacheExpiration()
+      );
+    }
+
+    return this.pluginCache.get(instanceId);
+  }
+
+  private emptyBodyFilter(
+    req: express.Request,
+    res: express.Response,
+    next: express.NextFunction
+  ) {
+    if (!req.body || _.isEmpty(req.body)) {
+      const msg = {
+        error: "Missing request body",
+      };
+      this.logger.error(`POST /v1/${req.url} : %s`, JSON.stringify(msg));
```

Comment:
Similar remarks than the one below -> use `%j` instead of `%s`+`JSON.stringify()`